File: app/src/main/java/io/github/tobyhs/weatherweight/AppModule.java
package io.github.tobyhs.weatherweight;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import javax.inject.Named;
import javax.inject.Singleton;

import com.github.tobyhs.rxsecretary.SchedulerProvider;
import com.github.tobyhs.rxsecretary.android.AndroidSchedulerProvider;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import dagger.Module;
import dagger.Provides;

import okhttp3.OkHttpClient;

import org.aaronhe.threetengson.ThreeTenGsonAdapter;

import org.threeten.bp.Clock;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherApiKeyInterceptor;
import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherService;
import io.github.tobyhs.weatherweight.data.AccuWeatherRepository;
import io.github.tobyhs.weatherweight.data.WeatherRepository;
import io.github.tobyhs.weatherweight.storage.LastForecastStore;
import io.github.tobyhs.weatherweight.storage.SharedPrefLastForecastStore;
import io.github.tobyhs.weatherweight.util.GVTypeAdapterFactory;

/**
 * Dagger module to provide common dependencies
 */
@Module
public class AppModule {
    private final App app;

    /**
     * @param app application instance
     */
    public AppModule(App app) {
        this.app = app;
    }

    /**
     * @return application instance
     */
    @Provides
    @Singleton
    App provideApp() {
        return app;
    }

    /**
     * @return {@link SharedPreferences} instance for the application
     */
    @Provides
    @Singleton
    SharedPreferences provideSharedPreferences() {
        return PreferenceManager.getDefaultSharedPreferences(app);
    }

    /**
     * @return RxJava scheduler provider
     */
    @Provides
    @Singleton
    SchedulerProvider provideSchedulerProvider() {
        return new AndroidSchedulerProvider();
    }

    /**
     * @return a Gson instance
     */
    @Provides
    @Singleton
    public static Gson provideGson() {
        GsonBuilder gsonBuilder = new GsonBuilder()
                .registerTypeAdapterFactory(GVTypeAdapterFactory.create());
        return ThreeTenGsonAdapter.registerAll(gsonBuilder).create();
    }

    /**
     * @param gson Gson instance for parsing JSON response bodies
     * @return a Retrofit instance for AccuWeather's APIs
     */
    @Provides @Named("accuWeatherRetrofit")
    @Singleton
    Retrofit provideAccuWeatherRetrofit(Gson gson) {
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(new AccuWeatherApiKeyInterceptor(BuildConfig.ACCUWEATHER_API_KEY))
                .build();
        return new Retrofit.Builder()
                .baseUrl("https://dataservice.accuweather.com/")
                .client(okHttpClient)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
    }

    /**
     * @param retrofit a Retrofit instance for AccuWeather's APIs
     * @return Retrofit service object for querying weather info from AccuWeather's APIs
     */
    @Provides
    @Singleton
    AccuWeatherService provideAccuWeatherService(@Named("accuWeatherRetrofit") Retrofit retrofit) {
        return retrofit.create(AccuWeatherService.class);
    }

    /**
     * @param service Retrofit service object for querying weather data from AccuWeather's APIs
     * @return repository object to fetch weather data
     */
    @Provides
    @Singleton
    WeatherRepository provideWeatherRepository(AccuWeatherService service) {
        return new AccuWeatherRepository(service, Clock.systemDefaultZone());
    }

    /**
     * @param sharedPreferences {@link SharedPreferences} instance for the application
     * @param gson Gson instance to serialize data
     * @return store to save or get the last forecast
     */
    @Provides
    @Singleton
    LastForecastStore provideLastForecastStore(SharedPreferences sharedPreferences, Gson gson) {
        return new SharedPrefLastForecastStore(sharedPreferences, gson);
    }
}

File: app/src/main/java/io/github/tobyhs/weatherweight/AppModule.kt
package io.github.tobyhs.weatherweight

import android.content.Context

import com.github.tobyhs.rxsecretary.SchedulerProvider
import com.github.tobyhs.rxsecretary.android.AndroidSchedulerProvider

import com.squareup.moshi.Moshi

import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent

import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherApiKeyInterceptor
import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherService
import io.github.tobyhs.weatherweight.data.AccuWeatherRepository
import io.github.tobyhs.weatherweight.data.WeatherRepository
import io.github.tobyhs.weatherweight.data.adapter.LocalDateAdapter
import io.github.tobyhs.weatherweight.data.adapter.ZonedDateTimeAdapter
import io.github.tobyhs.weatherweight.storage.FileLastForecastStore
import io.github.tobyhs.weatherweight.storage.LastForecastStore

import java.io.File
import java.time.Clock
import javax.inject.Named
import javax.inject.Singleton

import okhttp3.Cache
import okhttp3.OkHttpClient

import retrofit2.Retrofit
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory

/**
 * Dagger module to provide common dependencies
 */
@Module
@InstallIn(SingletonComponent::class)
class AppModule {
    /**
     * @return RxJava scheduler provider
     */
    @Provides
    @Singleton
    fun provideSchedulerProvider(): SchedulerProvider = AndroidSchedulerProvider()

    /**
     * @param context application context
     * @return an OkHttp client for AccuWeather's APIs
     */
    @Provides
    @Named("accuWeatherOkHttp")
    @Singleton
    fun provideAccuWeatherOkHttp(@ApplicationContext context: Context): OkHttpClient {
        return OkHttpClient.Builder()
            .addInterceptor(AccuWeatherApiKeyInterceptor(BuildConfig.ACCUWEATHER_API_KEY))
            .cache(Cache(File(context.cacheDir, "AccuWeather"), 96000))
            .build()
    }

    /**
     * @param okHttpClient HTTP client for AccuWeather's APIs
     * @param moshi Moshi instance for parsing JSON responses
     * @return a Retrofit instance for AccuWeather's APIs
     */
    @Provides
    @Named("accuWeatherRetrofit")
    @Singleton
    fun provideAccuWeatherRetrofit(
        @Named("accuWeatherOkHttp") okHttpClient: OkHttpClient,
        moshi: Moshi,
    ): Retrofit {
        return Retrofit.Builder()
            .baseUrl("https://dataservice.accuweather.com/")
            .client(okHttpClient)
            .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
            .addConverterFactory(MoshiConverterFactory.create(moshi))
            .build()
    }

    /**
     * @param retrofit a Retrofit instance for AccuWeather's APIs
     * @return Retrofit service object for querying weather info from AccuWeather's APIs
     */
    @Provides
    @Singleton
    fun provideAccuWeatherService(@Named("accuWeatherRetrofit") retrofit: Retrofit): AccuWeatherService {
        return retrofit.create(AccuWeatherService::class.java)
    }

    /**
     * @param service Retrofit service object for querying weather data from AccuWeather's APIs
     * @return repository object to fetch weather data
     */
    @Provides
    @Singleton
    fun provideWeatherRepository(service: AccuWeatherService): WeatherRepository {
        return AccuWeatherRepository(service, Clock.systemDefaultZone())
    }

    /**
     * @param context application context
     * @param moshi Moshi instance to serialize data
     * @return store to save or get the last forecast
     */
    @Provides
    @Singleton
    fun provideLastForecastStore(@ApplicationContext context: Context, moshi: Moshi): LastForecastStore {
        return FileLastForecastStore(context.cacheDir, moshi)
    }

    companion object {
        /**
         * @return a Moshi instance
         */
        @Provides
        @Singleton
        fun provideMoshi(): Moshi = Moshi.Builder()
            .add(LocalDateAdapter())
            .add(ZonedDateTimeAdapter())
            .build()
    }
}
